src/lib/browse/Filters.svelte

```diff
@@ -0,0 +1,127 @@
+<script lang="ts">
+  import { onMount } from "svelte";
+  import { gjScheme } from "../../stores";
+  import CollapsibleCard from "../common/CollapsibleCard.svelte";
+  import FormElement from "../govuk/FormElement.svelte";
+  import SecondaryButton from "../govuk/SecondaryButton.svelte";
+  import Select from "../govuk/Select.svelte";
+  import { type Scheme } from "./data";
+
+  // These are immutable; re-create this component if they change
+  export let schemes: Map<string, Scheme>;
+
+  // by scheme_reference
+  export let showSchemes: Set<string> = new Set();
+
+  let filterText = "";
+
+  // Dropdown filters
+  let authorities: [string, string][] = [];
+  let filterAuthority = "";
+  let fundingProgrammes: [string, string][] = [];
+  let filterFundingProgramme = "";
+
+  // Stats about filtered schemes
+  let counts = { area: 0, route: 0, crossing: 0, other: 0 };
+
+  onMount(() => {
+    let set1: Set<string> = new Set();
+    let set2: Set<string> = new Set();
+    for (let x of schemes.values()) {
+      if (x.authority_or_region) {
+        set1.add(x.authority_or_region);
+      }
+      if (x.funding_programme) {
+        set2.add(x.funding_programme);
+      }
+    }
+    authorities = Array.from(set1.entries());
+    authorities.sort();
+    fundingProgrammes = Array.from(set2.entries());
+    fundingProgrammes.sort();
+  });
+
+  // When any filters change, update showSchemes
+  function filtersUpdated(
+    filterText: string,
+    filterAuthority: string,
+    filterFundingProgramme: string
+  ) {
+    showSchemes.clear();
+    let filterNormalized = filterText.toLowerCase();
+    for (let feature of $gjScheme.features) {
```

Comment:
Can we use an actual `Array.filter`?
Something like: `schemesToBeShown = schemes.filter(getPredicate(filterText, filterAuthority, filterFundingProgramme))`

If we want to make it more performant we store previous filters somewhere and work out how our filters have changed and if we are simply getting more specific (e.g. filter authority stays the same but we add text or funding programme where we didn't have that before) in which case we can filter on the existing schemes being shown.

Reply:
Changed to use filter and map. I love the idea of filtering on prior results when the only change is appending to filterText -- that's a comm one! I haven't yet profiled to figure out where the (not so terrible) lag in filtering comes from yet -- maybe recalculating the set is fine, and updating maplibre is the slow part.